src/Compiler/Service/ServiceParseTreeWalk.fs

```diff
@@ -1103,9 +1114,478 @@ module SyntaxTraversal =
 
             loop diveResults
 
-        ignore<'T option> (traverseUntil pick parseTree.Range.End visitor parseTree)
+        let contents =
+            match parseTree with
+            | ParsedInput.ImplFile implFile -> implFile.Contents |> List.map SyntaxNode.SynModuleOrNamespace
+            | ParsedInput.SigFile sigFile -> sigFile.Contents |> List.map SyntaxNode.SynModuleOrNamespaceSig
+
+        ignore<'T option> (traverseUntil pick parseTree.Range.End visitor contents)
 
     /// traverse an implementation file walking all the way down to SynExpr or TypeAbbrev at a particular location
     ///
     let Traverse (pos: pos, parseTree, visitor: SyntaxVisitorBase<'T>) =
-        traverseUntil pick pos visitor parseTree
+        let contents =
+            match parseTree with
+            | ParsedInput.ImplFile implFile -> implFile.Contents |> List.map SyntaxNode.SynModuleOrNamespace
+            | ParsedInput.SigFile sigFile -> sigFile.Contents |> List.map SyntaxNode.SynModuleOrNamespaceSig
+
+        traverseUntil pick pos visitor contents
+
+[<RequireQualifiedAccess>]
+[<CompilationRepresentation(CompilationRepresentationFlags.ModuleSuffix)>]
+module SyntaxNode =
+    let (|Attributes|) node =
+        let (|All|) = List.collect
+        let field (SynField(attributes = attributes)) = attributes
+        let unionCase (SynUnionCase(attributes = attributes)) = attributes
+        let enumCase (SynEnumCase(attributes = attributes)) = attributes
+        let typar (SynTyparDecl(attributes = attributes)) = attributes
+
+        let (|SynComponentInfo|) =
+            function
```

Comment:
Not for this PR, but I think in compiler we should really prefer explicit matches, instead of "function" shorthand, matches are easier to debug and see explicit bindings.

Reply:
12ae786